src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1188,30 +1192,88 @@ static int32_t CopyFile_ReadWrite(int inFd, int outFd)
 }
 #endif // !HAVE_FCOPYFILE
 
-int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
+int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char* destPath, int32_t overwrite)
 {
     int inFd = ToFileDescriptor(sourceFd);
-    int outFd = ToFileDescriptor(destinationFd);
-
-#if HAVE_FCOPYFILE
-    // If fcopyfile is available (OS X), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
-    // Copy data and metadata.
-    return fcopyfile(inFd, outFd, NULL, COPYFILE_ALL) == 0 ? 0 : -1;
-#else
-    // Get the stats on the source file.
+    int outFd;
     int ret;
     struct stat_ sourceStat;
-    bool copied = false;
-#if HAVE_SENDFILE_4
-    // If sendfile is available (Linux), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
+
     while ((ret = fstat_(inFd, &sourceStat)) < 0 && errno == EINTR);
     if (ret != 0)
     {
         return -1;
     }
 
+    struct stat_ destStat;
+    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
+    if (ret == 0)
+    {
+        if (sourceStat.st_dev == destStat.st_dev && sourceStat.st_ino == destStat.st_ino)
+        {
+            errno = EBUSY;
+            return -1;
+        }
+
+        if (!overwrite)
+        {
+            errno = EEXIST;
+            return -1;
+        }
+
+#if HAVE_CLONEFILE
+        // For clonefile we need to unlink the destination file first but we need to
+        // check permission first to ensure we don't try to unlink read-only file
+        if (access(destPath, W_OK) != 0)
+        {
+            errno = EACCES;
+            return -1;
+        }
+        
+        ret = unlink(destPath);
+        if (ret != 0)
+        {
+            return ret;
+        }
+#endif
+    }
+
+#if HAVE_CLONEFILE
+    while ((ret = clonefile(srcPath, destPath, 0)) < 0 && errno == EINTR);
+    // EEXIST can happen due to race condition between the stat/unlink above
+    // and the clonefile here. The file could be (re-)created from another
+    // thread or process before we have a chance to call clonefile. Handle
+    // it by falling back to the slow path.
+    if (ret == 0 || (errno != ENOTSUP && errno != EXDEV && errno != EEXIST))
+    {
+        return ret;
+    }
+#else
+    // Unused variable
+    (void)srcPath;
+#endif
+
+    while ((outFd = open(destPath, O_WRONLY | O_TRUNC | O_CREAT | (overwrite ? 0 : O_EXCL), sourceStat.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO))) < 0 && errno == EINTR);
+    if (outFd < 0)
+    {
+        return -1;
+    }
+
+#if HAVE_FCOPYFILE
+    // If fcopyfile is available (macOS), try to use it, as it handles
+    // copying both data and metadata. Unlike sendfile below it's
+    // implemented as user space library but future version of macOS
+    // may optimize it.
+    ret = fcopyfile(inFd, outFd, NULL, COPYFILE_ALL) == 0 ? 0 : -1;
```

Comment:
From your comments on the mono issue/PR, it sounds like you were saying that fcopyfile isn't actually better than the same read/write loop we do below.  If that's true, can we just delete this HAVE_FCOPYFILE code?  Or are there still advantages to it?

In particular, it seems like we'd only end up here in a rare situation where clonefile failed; even if fcopyfile is slightly better than our own read/write loop, not sure it's worth the extra complexity for that rare fallback case.

But maybe I'm misunderstanding?

Reply:
I'd be happy to remove it. My only concern was if macOS added some metadata that would be copied by `fcopyfile` but not by the CoreFX code. The source code for [the copyfile package](https://opensource.apple.com/source/copyfile/copyfile-138/copyfile.c.auto.html) is available though and it does not seem to be the case.